### PR TITLE
Fix polling transports and add tests for closing transports

### PR DESCRIPTION
**Pull request by nkzawa**
Added tests for https://github.com/socketio/engine.io/pull/344, https://github.com/socketio/engine.io/pull/340, https://github.com/socketio/engine.io/pull/339 and https://github.com/socketio/engine.io/pull/336.

I have to fix some codes for writing these tests.
- add `upgrading` event on socket
- fix closing polling transports

### CODE REVIEW

File: lib/socket.js [hunk 1]
```diff
@@ -93,7 +93,6 @@ Socket.prototype.onPacket = function (packet) {
         break;
 
       case 'error':
-        this.transport.close();
```
Comment (nkzawa) [inline lib/socket.js]: The reason of socket closing becomes "transport close" if we call here.
This method will be called on `clearTransport` anyway.

File: lib/transports/polling.js [hunk 2]
```diff
@@ -372,9 +377,16 @@ Polling.prototype.doClose = function (fn) {
...
     debug('transport not writable - buffering orderly close');
-    this.shouldClose = fn;
+    this.shouldClose = onClose;
+    closeTimeoutTimer = setTimeout(onClose, this.closeTimeout);
+  }
+
+  function onClose() {
+    clearTimeout(closeTimeoutTimer);
+    fn();
+    self.onClose();
```
Comment (nkzawa) [inline lib/transports/polling.js]: `onClose` would never be called if we didn't manually call it.

File: lib/transports/polling.js [hunk 3]
```diff
@@ -372,9 +377,16 @@ Polling.prototype.doClose = function (fn) {
   if (this.writable) {
     debug('transport writable - closing right away');
     this.send([{ type: 'close', options: { compress: true } }]);
-    fn();
+    onClose();
   } else {
     debug('transport not writable - buffering orderly close');
-    this.shouldClose = fn;
+    this.shouldClose = onClose;
+    closeTimeoutTimer = setTimeout(onClose, this.closeTimeout);
```
Comment (nkzawa) [inline lib/transports/polling.js]: for the case that clients didn't send a poll request anymore.

**Comment by Mumakil** [comment]
We tried running this, but ended up in a situation where the websocket disconnects 30s after an upgrade from xhr-polling. My initial take would be that the event listeners from the old transport are not cleared when the client upgrades and the closeTimeout fires and emits a `close` event.

**Comment by nkzawa** [comment]
Thanks for your report! I will check the problem.

**Comment by nkzawa** [comment]
The problem should be fixed now.

**Comment by nicofff** [comment]
Just tested this patch. It's causing reconnects every 30s

**Comment by nkzawa** [comment]
@nicofff I just tested with socket.io but it was ok. How can I reproduce?

**Pull request merged**